In [1]:
import pandas as pd
import json

# Narrowing down to one subset of diseases

In [2]:
DATASET_PATH = '/Users/rebeccaceppas/Desktop/code/mywork/project/dataset/'

In [3]:
with open(DATASET_PATH+'release_conditions.json') as f:
   pc = json.load(f)
f.close()

possible_conditions = pd.DataFrame.from_dict(pc)

In [4]:
# Immune and Infectious Diseases

# Infectious Diseases
infectious_diseases = [
    'VIH (Primo-infection)',
    'Coqueluche',
    'Chagas',
    'Tuberculose',
    'Ebola',
    'Possible influenza ou syndrome virémique typique'
]

# Autoimmune and Inflammatory Conditions
autoimmune_and_inflammatory_conditions = [
    'Lupus érythémateux disséminé (LED)',
    'Sarcoïdose'
]

# Allergic Reactions and Related Conditions
allergic_reactions_and_related_conditions = [
    'Anaphylaxie',
    'Rhinite allergique',
    'Oedème localisé ou généralisé sans atteinte pulmonaire associée'
]

immune_infectious = infectious_diseases + autoimmune_and_inflammatory_conditions + allergic_reactions_and_related_conditions

# Making datasets only including patients with these diseases

In [5]:
train = pd.read_csv(DATASET_PATH+'release_train_patients.csv')
val = pd.read_csv(DATASET_PATH+'release_validate_patients.csv')
test = pd.read_csv(DATASET_PATH+'release_test_patients.csv')

In [6]:
train_patients = []
val_patients = []
test_patients = []

for ii in immune_infectious:
    train_patients += [ind for ind in train.PATHOLOGY[train.PATHOLOGY == ii].index.values]
    val_patients += [ind for ind in val.PATHOLOGY[val.PATHOLOGY == ii].index.values]
    test_patients += [ind for ind in test.PATHOLOGY[test.PATHOLOGY == ii].index.values]


In [7]:
# selecting new samples of patients
train_ii = train.iloc[train_patients]
valid_ii = val.iloc[val_patients]
test_ii = test.iloc[test_patients]

In [8]:
print('There are now', len(train_ii), 'patients in the training set.')
print('There are now', len(valid_ii), 'patients in the validation set.')
print('There are now', len(test_ii), 'patients in the testing set.')

There are now 203008 patients in the training set.
There are now 25399 patients in the validation set.
There are now 25689 patients in the testing set.


In [9]:
test_ii.head()

AGE                             DIFFERENTIAL_DIAGNOSIS SEX  \
46    17  [['VIH (Primo-infection)', 0.3903549994602378]...   M   
114   35  [['VIH (Primo-infection)', 0.339660592404885],...   M   
122   15  [['VIH (Primo-infection)', 0.3925425839871151]...   F   
123    4  [['VIH (Primo-infection)', 0.441221802327311],...   F   
154    0  [['VIH (Primo-infection)', 0.43630678617344326...   M   

                 PATHOLOGY                                          EVIDENCES  \
46   VIH (Primo-infection)  ['adp_dlr', 'atcd_its', 'diarrhee', 'douleurxx...   
114  VIH (Primo-infection)  ['adp_dlr', 'atcd_its', 'diaph', 'diarrhee', '...   
122  VIH (Primo-infection)  ['adp_dlr', 'atcd_its', 'diaph', 'douleurxx', ...   
123  VIH (Primo-infection)  ['adp_dlr', 'diarrhee', 'douleurxx', 'douleurx...   
154  VIH (Primo-infection)  ['adp_dlr', 'diaph', 'diarrhee', 'douleurxx', ...   

    INITIAL_EVIDENCE  
46         douleurxx  
114        fatig_ext  
122            diaph  
123           fievre  
154          adp_dlr

# Saving the data as new csv files

In [10]:
train_ii.to_csv(DATASET_PATH+'subset_train.csv', index=False)
valid_ii.to_csv(DATASET_PATH+'subset_validation.csv', index=False)
test_ii.to_csv(DATASET_PATH+'subset_test.csv', index=False)

In [12]:
test_test = pd.read_csv(DATASET_PATH+'subset_test.csv')

In [13]:
test_test.head()

AGE                             DIFFERENTIAL_DIAGNOSIS SEX  \
0   17  [['VIH (Primo-infection)', 0.3903549994602378]...   M   
1   35  [['VIH (Primo-infection)', 0.339660592404885],...   M   
2   15  [['VIH (Primo-infection)', 0.3925425839871151]...   F   
3    4  [['VIH (Primo-infection)', 0.441221802327311],...   F   
4    0  [['VIH (Primo-infection)', 0.43630678617344326...   M   

               PATHOLOGY                                          EVIDENCES  \
0  VIH (Primo-infection)  ['adp_dlr', 'atcd_its', 'diarrhee', 'douleurxx...   
1  VIH (Primo-infection)  ['adp_dlr', 'atcd_its', 'diaph', 'diarrhee', '...   
2  VIH (Primo-infection)  ['adp_dlr', 'atcd_its', 'diaph', 'douleurxx', ...   
3  VIH (Primo-infection)  ['adp_dlr', 'diarrhee', 'douleurxx', 'douleurx...   
4  VIH (Primo-infection)  ['adp_dlr', 'diaph', 'diarrhee', 'douleurxx', ...   

  INITIAL_EVIDENCE  
0        douleurxx  
1        fatig_ext  
2            diaph  
3           fievre  
4          adp_dlr

# Looking at the evidences files

In [18]:
with open(DATASET_PATH+'release_evidences.json') as f:
   pe = json.load(f)
f.close()

possible_evidences = pd.DataFrame.from_dict(pe).T

In [24]:
possible_evidences.info()

<class 'pandas.core.frame.DataFrame'>
Index: 223 entries, fievre to J81
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   name             223 non-null    object
 1   code_question    223 non-null    object
 2   question_fr      223 non-null    object
 3   question_en      223 non-null    object
 4   is_antecedent    223 non-null    object
 5   default_value    223 non-null    object
 6   value_meaning    223 non-null    object
 7   possible-values  223 non-null    object
 8   data_type        223 non-null    object
dtypes: object(9)
memory usage: 17.4+ KB


In [32]:
possible_evidences.sample(15)

name   code_question  \
ww_respi                              ww_respi        ww_respi   
dysp_effort                        dysp_effort     dysp_effort   
k21                                        k21             k21   
vo_violent                          vo_violent      vo_violent   
wheez                                    wheez           wheez   
hernie_hiatale                  hernie_hiatale  hernie_hiatale   
synd_nephro                        synd_nephro     synd_nephro   
obstipation                        obstipation     obstipation   
B34.9                                    B34.9           B34.9   
trav1                                    trav1           trav1   
e78.5                                    e78.5           e78.5   
douleurxx_soudain            douleurxx_soudain       douleurxx   
rds_sg                                  rds_sg          rds_sg   
lesions_peau_plusqu1cm  lesions_peau_plusqu1cm    lesions_peau   
douleurxx_irrad                douleurxx_irrad       douleurxx   

                                                              question_fr  \
ww_respi                Avez-vous une douleur qui est pire à l’inspira...   
dysp_effort             Vous sentez-vous facilement essoufflé à l’effort?   
k21                     Avez-vous déjà reçu un diagnostic de reflux ga...   
vo_violent              Avez-vous eu plusieurs vomissements ou fait pl...   
wheez                   Avez-vous noté un un son aigu lorsque vous sor...   
hernie_hiatale                              Avez-vous une hernie hiatale?   
synd_nephro             Avez-vous un problème rénal qui se caractérise...   
obstipation             Passez-vous des selles ou des gaz depuis que v...   
B34.9                       Avez-vous été infecté par un virus récemment?   
trav1                     Avez-vous voyagé dans les 4 dernières semaines?   
e78.5                   Souffrez-vous d’hypercholestérolémie ou prenez...   
douleurxx_soudain          À quelle vitesse la douleur est-elle apparue ?   
rds_sg                  Avez-vous remarqué des saignement ou des bleus...   
lesions_peau_plusqu1cm  Est-ce que la ou les lésions sont plus grande ...   
douleurxx_irrad         Est-ce que la douleur se propage vers un autre...   

                                                              question_en  \
ww_respi                Do you have pain that is increased when you br...   
dysp_effort             Do you feel out of breath with minimal physica...   
k21                     Have you ever been diagnosed with gastroesopha...   
vo_violent              Have you vomited several times or have you mad...   
wheez                   Have you noticed a wheezing sound when you exh...   
hernie_hiatale                               Do you have a hiatal hernia?   
synd_nephro             Do you have a known kidney problem resulting i...   
obstipation             Have you been able to pass stools or gas since...   
B34.9                            Have you recently had a viral infection?   
trav1                   Have you traveled out of the country in the la...   
e78.5                   Do you have high cholesterol or do you take me...   
douleurxx_soudain                           How fast did the pain appear?   
rds_sg                  Have you noticed any unusual bleeding or bruis...   
lesions_peau_plusqu1cm  Is the lesion (or are the lesions) larger than...   
douleurxx_irrad                Does the pain radiate to another location?   

                       is_antecedent default_value  \
ww_respi                       False             0   
dysp_effort                    False             0   
k21                             True             0   
vo_violent                     False             0   
wheez                          False             0   
hernie_hiatale                  True             0   
synd_nephro                     True             0   
obstipation                    False             0   
B34.9                           T

In [37]:
possible_evidences.loc['douleurxx_irrad']['possible-values']

['nulle_part',
 'aile_iliaque_D_',
 'aile_iliaque_G_',
 'aine_D_',
 'aine_G_',
 'aisselle_D_',
 'aisselle_G_',
 'amygdale_D_',
 'amygdale_G_',
 'anus',
 'arrière_de_la_cheville_D_',
 'arrière_de_la_cheville_G_',
 'arrière_de_tête',
 'arrière_du_cou',
 'avant-bras_D_',
 'avant-bras_G_',
 'bas_du_thorax',
 'biceps_D_',
 'biceps_G_',
 'bouche',
 'cartilage_thyroidien',
 'cheville_D_',
 'cheville_G_',
 'clitoris',
 'coccyx',
 'colonne_cervicale',
 'colonne_dorsale',
 'colonne_lombaire',
 'commissure_D_',
 'commissure_G_',
 'coté_lateral_du_pied_D_',
 'coté_lateral_du_pied_G_',
 'coude_D_',
 'coude_G_',
 'creux_poplité_D_',
 'creux_poplité_G_',
 'crête_iliaque_D_',
 'crête_iliaque_G_',
 'cuisse_D_',
 'cuisse_G_',
 'côté_du_cou_D_',
 'côté_du_cou_G_',
 'côté_du_thorax_D_',
 'côté_du_thorax_G_',
 'dents_inférieures_D_',
 'dents_inférieures_G_',
 'dents_supérieures_D_',
 'dents_supérieures_G_',
 'dessus_de_la_langue',
 'dessus_de_tête',
 'doigt_annulaire__D_',
 'doigt_annulaire__G_',
 'doigt_a

In [42]:
len(set(possible_evidences.loc['douleurxx_irrad']['possible-values']))

165

#### Note: we found that there are 223 different symptoms and for the multi-choice evidences there would be multiple columns for each containing 0 or 1 for selecting one of the choices. If we have some multi-choice options with a lot of options (like the one above with 165 options), this could get out of hand.

#### Posisble idea: make it so that all multi-choice options are now binary so that they are 1 of any of the options are selected and 0 if none of them are.

# Looking at categorical features

For the ones with numbers, leave the columns as it is. For N,O questions, keep binary. For skin rash color make 5 separate columns for 6 options. For travel out of country, make it binary.

In [59]:
possible_evidences[possible_evidences.data_type == 'C']

name code_question  \
douleurxx_soudain            douleurxx_soudain     douleurxx   
douleurxx_intens              douleurxx_intens     douleurxx   
douleurxx_precis              douleurxx_precis     douleurxx   
lesions_peau_couleur      lesions_peau_couleur  lesions_peau   
lesions_peau_intens        lesions_peau_intens  lesions_peau   
lesions_peau_elevee        lesions_peau_elevee  lesions_peau   
lesions_peau_prurit        lesions_peau_prurit  lesions_peau   
lesions_peau_plusqu1cm  lesions_peau_plusqu1cm  lesions_peau   
lesions_peau_desquame    lesions_peau_desquame  lesions_peau   
trav1                                    trav1         trav1   

                                                              question_fr  \
douleurxx_soudain          À quelle vitesse la douleur est-elle apparue ?   
douleurxx_intens                    Quelle est l’intensité de la douleur?   
douleurxx_precis        À quel point la douleur est-elle précisément l...   
lesions_peau_couleur                  De quelle couleur sont les lésions?   
lesions_peau_intens     Quelle est l’intensité de la douleur causée pa...   
lesions_peau_elevee              La ou les lésions sont-elles surélevées?   
lesions_peau_prurit               Quel est le niveau de la démangeaisons?   
lesions_peau_plusqu1cm  Est-ce que la ou les lésions sont plus grande ...   
lesions_peau_desquame   Est-ce que vos lésions pèlent (morceaux de pea...   
trav1                     Avez-vous voyagé dans les 4 dernières semaines?   

                                                              question_en  \
douleurxx_soudain                           How fast did the pain appear?   
douleurxx_intens                                 How intense is the pain?   
douleurxx_precis                       How precisely is the pain located?   
lesions_peau_couleur                              What color is the rash?   
lesions_peau_intens           How intense is the pain caused by the rash?   
lesions_peau_elevee                                  Is the rash swollen?   
lesions_peau_prurit                            How severe is the itching?   
lesions_peau_plusqu1cm  Is the lesion (or are the lesions) larger than...   
lesions_peau_desquame                           Do your lesions peel off?   
trav1                   Have you traveled out of the country in the la...   

                       is_antecedent default_value  \
douleurxx_soudain              False             0   
douleurxx_intens               False             0   
douleurxx_precis               False             0   
lesions_peau_couleur           False            NA   
lesions_peau_intens            False             0   
lesions_peau_elevee            False             0   
lesions_peau_prurit            False             0   
lesions_peau_plusqu1cm         False             N   
lesions_peau_desquame          False             N   
trav1                           True             N   

                                                            value_meaning  \
douleurxx_soudain                                                      {}   
douleurxx_intens                                                       {}   
douleurxx_precis                                                       {}   
lesions_peau_couleur    {'NA': {'fr': 'NA', 'en': 'NA'}, 'foncee': {'f...   
lesions_peau_intens                                                    {}   
lesions_peau_elevee                                                    {}   
lesions_peau_prurit                                                    {}   
lesions_peau_plusqu1cm  {'N': {'fr': 'N', 'en': 'N'}, 'O': {'fr': 'O',...   
lesions_peau_desquame   {'N': {'fr': 'N', 'en': 'N'}, 'O': {'fr': 'O',...   
trav1                   {'N': {'fr': 'N', 'en': 'N'}, 'AfriqN': {'fr':...   

                                                          possible-values  \
douleurxx_soudain                      [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]   
douleurxx_intens                       [0, 1

# Looking at multi-choice features

Group the body parts so we have less columns.

In [60]:
possible_evidences[possible_evidences.data_type == 'M']


name code_question  \
douleurxx_endroitducorps        douleurxx_endroitducorps     douleurxx   
douleurxx_irrad                          douleurxx_irrad     douleurxx   
douleurxx_carac                          douleurxx_carac     douleurxx   
lesions_peau_endroitducorps  lesions_peau_endroitducorps  lesions_peau   
oedeme_endroitducorps              oedeme_endroitducorps        oedeme   

                                                                   question_fr  \
douleurxx_endroitducorps                 Avez-vous de la douleur quelque part?   
douleurxx_irrad              Est-ce que la douleur se propage vers un autre...   
douleurxx_carac                                    Caractérisez votre douleur:   
lesions_peau_endroitducorps            À quel endroit est situé chaque lésion?   
oedeme_endroitducorps                      À quel endroit est situé l’enflure?   

                                                            question_en  \
douleurxx_endroitducorps                    Do you feel pain somewhere?   
douleurxx_irrad              Does the pain radiate to another location?   
douleurxx_carac                                 Characterize your pain:   
lesions_peau_endroitducorps       Where is the affected region located?   
oedeme_endroitducorps                    Where is the swelling located?   

                            is_antecedent default_value  \
douleurxx_endroitducorps            False    nulle_part   
douleurxx_irrad                     False    nulle_part   
douleurxx_carac                     False            NA   
lesions_peau_endroitducorps         False    nulle_part   
oedeme_endroitducorps               False    nulle_part   

                                                                 value_meaning  \
douleurxx_endroitducorps     {'nulle_part': {'fr': 'nulle part', 'en': 'now...   
douleurxx_irrad              {'nulle_part': {'fr': 'nulle part', 'en': 'now...   
douleurxx_carac              {'NA': {'fr': 'NA', 'en': 'NA'}, 'déchirante':...   
lesions_peau_endroitducorps  {'nulle_part': {'fr': 'nulle part', 'en': 'now...   
oedeme_endroitducorps        {'nulle_part': {'fr': 'nulle part', 'en': 'now...   

                                                               possible-values  \
douleurxx_endroitducorps     [nulle_part, aile_iliaque_D_, aile_iliaque_G_,...   
douleurxx_irrad              [nulle_part, aile_iliaque_D_, aile_iliaque_G_,...   
douleurxx_carac              [NA, déchirante, lancinante_/_choc_électrique,...   
lesions_peau_endroitducorps  [nulle_part, aile_iliaque_D_, aile_iliaque_G_,...   
oedeme_endroitducorps        [nulle_part, aile_iliaque_D_, aile_iliaque_G_,...   

                            data_type  
douleurxx_endroitducorps            M  
douleurxx_irrad                     M  
douleurxx_carac                     M  
lesions_peau_endroitducorps         M  
oedeme_endroitducorps               M

In [81]:
possible_evidences[possible_evidences.data_type == "M"]['possible-values'].iloc[0]

['nulle_part',
 'aile_iliaque_D_',
 'aile_iliaque_G_',
 'aine_D_',
 'aine_G_',
 'aisselle_D_',
 'aisselle_G_',
 'amygdale_D_',
 'amygdale_G_',
 'anus',
 'arrière_de_la_cheville_D_',
 'arrière_de_la_cheville_G_',
 'arrière_de_tête',
 'arrière_du_cou',
 'avant-bras_D_',
 'avant-bras_G_',
 'bas_du_thorax',
 'biceps_D_',
 'biceps_G_',
 'bouche',
 'cartilage_thyroidien',
 'cheville_D_',
 'cheville_G_',
 'clitoris',
 'coccyx',
 'colonne_cervicale',
 'colonne_dorsale',
 'colonne_lombaire',
 'commissure_D_',
 'commissure_G_',
 'coté_lateral_du_pied_D_',
 'coté_lateral_du_pied_G_',
 'coude_D_',
 'coude_G_',
 'creux_poplité_D_',
 'creux_poplité_G_',
 'crête_iliaque_D_',
 'crête_iliaque_G_',
 'cuisse_D_',
 'cuisse_G_',
 'côté_du_cou_D_',
 'côté_du_cou_G_',
 'côté_du_thorax_D_',
 'côté_du_thorax_G_',
 'dents_inférieures_D_',
 'dents_inférieures_G_',
 'dents_supérieures_D_',
 'dents_supérieures_G_',
 'dessus_de_la_langue',
 'dessus_de_tête',
 'doigt_annulaire__D_',
 'doigt_annulaire__G_',
 'doigt_a

In [84]:
body_parts_dict = {
    'head': ['arrière_de_tête', 'dessus_de_tête', 'front', 'menton', 'occiput', 'tempe_D_', 'tempe_G_'],
    'face': ['oreille_D_', 'oreille_G_', 'oeil_D_', 'oeil_G_', 'nez', 'lèvre_inferieure_D_', 'lèvre_supérieure_D_', 'joue_D_', 'joue_G_', 'vermillon_D_', 'vermillon_G_'],
    'mouth': ['bouche', 'dents_inférieures_D_', 'dents_inférieures_G_', 'dents_supérieures_D_', 'dents_supérieures_G_', 'gencive_inférieure', 'gencive_supérieure', 'luette', 'palais', 'sous_la_langue', 'joue_interne_D_', 'joue_interne_G_', 'amygdale_D_', 'amygdale_G_'],
    'neck_throat': ['arrière_du_cou', 'côté_du_cou_D_', 'côté_du_cou_G_', 'cartilage_thyroidien', 'pharynx', 'trachée'],
    'thorax': ['bas_du_thorax', 'côté_du_thorax_D_', 'côté_du_thorax_G_', 'haut_du_thorax', 'thorax_postérieur_D_', 'thorax_postérieur_G_', 'ventre', 'épigastre', 'flanc_D_', 'flanc_G_', 'hypochondre_D_', 'hypochondre_G_'],
    'pelvis_hips': ['aine_D_', 'aine_G_', 'fosse_iliaque_D_', 'fosse_iliaque_G_', 'crête_iliaque_D_', 'crête_iliaque_G_', 'hanche_D_', 'hanche_G_', 'coccyx', 'aile_iliaque_D_', 'aile_iliaque_G_'],
    'upper_limb': ['avant-bras_D_', 'avant-bras_G_', 'biceps_D_', 'biceps_G_', 'coude_D_', 'coude_G_', 'doigt_annulaire__D_', 'doigt_annulaire__G_', 'doigt_auriculaire__D_', 'doigt_auriculaire__G_', 'doigt_index__D_', 'doigt_index__G_', 'doigt_majeur__D_', 'doigt_majeur__G_', 'épaule_D_', 'épaule_G_', 'face_dorsale_du_poignet_D_', 'face_dorsale_du_poignet_G_', 'face_palmaire_de_l_avant-bras_D_', 'face_palmaire_de_l_avant-bras_G_', 'face_palmaire_du_poignet_D_', 'face_palmaire_du_poignet_G_', 'paume_D_', 'paume_G_', 'pouce_D_', 'pouce_G_', 'trapèze_D_', 'trapèze_G_', 'triceps_D_', 'triceps_G_', 'aisselle_D_', 'aisselle_G_'],
    'lower_limb': ['cheville_D_', 'cheville_G_', 'cuisse_D_', 'cuisse_G_', 'genou_D_', 'genou_G_', 'ischio_D_', 'ischio_G_', 'mollet_D_', 'mollet_G_', 'pied_D_', 'pied_G_', 'gros_orteil_D_', 'gros_orteil_G_', 'orteil__1__D_', 'orteil__1__G_', 'orteil__2__D_', 'orteil__2__G_', 'orteil__3__D_', 'orteil__3__G_', 'petit_orteil__4__D_', 'petit_orteil__4__G_', 'talon_D_', 'talon_G_', 'fesse_D_', 'fesse_G_'],
    'back': ['colonne_cervicale', 'colonne_dorsale', 'colonne_lombaire'],
    'organs': ['clitoris', 'hymen', 'paroi_vaginale_D_', 'paroi_vaginale_G_', 'vagin', 'vestibule', 'pénis', 'scrotum', 'testicule_D_', 'testicule_G_', 'loges_renales_D_', 'loges_renales_G_', 'anus'],
}
